cvat/apps/dataset_manager/util.py

```diff
@@ -80,3 +92,140 @@ def deepcopy_simple(v):
         return v
     else:
         return deepcopy(v)
+
+
+class LockNotAvailableError(Exception):
+    pass
+
+
+def make_export_cache_lock_key(filename: os.PathLike[str]) -> str:
+    return f"export_lock:{filename}"
+
+
+@contextmanager
+def get_export_cache_lock(
+    export_path: os.PathLike[str],
+    *,
+    ttl: int | timedelta,
+    block: bool = True,
+    acquire_timeout: Optional[int | timedelta] = None,
+) -> Generator[Lock, Any, Any]:
+    if isinstance(acquire_timeout, timedelta):
+        acquire_timeout = acquire_timeout.total_seconds()
+    if acquire_timeout is not None and acquire_timeout < 0:
+        raise ValueError("acquire_timeout must be a non-negative number")
+    elif acquire_timeout is None:
+        acquire_timeout = -1
+
+    if isinstance(ttl, timedelta):
+        ttl = ttl.total_seconds()
+    if not ttl or ttl < 0:
+        raise ValueError("ttl must be a non-negative number")
+
+    # https://redis.io/docs/latest/develop/use/patterns/distributed-locks/
+    # The lock is exclusive, so it may potentially reduce performance in some cases,
+    # where parallel access is potentially possible and valid,
+    # e.g. dataset downloading could use a shared lock instead.
+    lock = Redlock(
+        key=make_export_cache_lock_key(export_path),
+        masters={django_rq.get_connection(settings.CVAT_QUEUES.EXPORT_DATA.value)},
+        auto_release_time=ttl,
+    )
+    try:
+        acquired = lock.acquire(blocking=block, timeout=acquire_timeout)
+        if acquired:
+            yield lock
+        else:
+            raise LockNotAvailableError
+
+    finally:
+        if lock.locked():
+            lock.release()
+
+
+EXPORT_CACHE_DIR_NAME = 'export_cache'
+
+
+def get_export_cache_dir(db_instance: Project | Task | Job) -> str:
+    base_dir = osp.abspath(db_instance.get_dirname())
+
+    if osp.isdir(base_dir):
+        return osp.join(base_dir, EXPORT_CACHE_DIR_NAME)
+    else:
+        raise FileNotFoundError(
+            '{} dir {} does not exist'.format(db_instance.__class__.__name__, base_dir)
+        )
+
+
+def make_export_filename(
+    dst_dir: str,
+    save_images: bool,
+    instance_update_time: datetime,
+    format_name: str,
+) -> str:
+    from .formats.registry import EXPORT_FORMATS
+    file_ext = EXPORT_FORMATS[format_name].EXT
+
+    filename = '%s-instance%s-%s.%s' % (
```

Comment:
You should use the fixed width format for the timestamp, otherwise, Python might choose exponential notation and the parsing will break.

Or you could round it to an int; it doesn't seem like you need sub-second precision here.

Reply:
Updated to use `%f`.